File: node/file_analysis.js
var fs = require('fs');
var readline = require('readline');
var path = require('path');
var yaml = require('js-yaml');

const Waiter = require('./yibu').Waiter;

var _tree = new Map();
var extensionCnt = new Map();
var extensionLineCnt = new Map();

var config = yaml.safeLoad(fs.readFileSync('config.yaml'));

let waiter = new Waiter();

config.extensions.forEach(function(value) {
    extensionCnt.set(value, 0);
    extensionLineCnt.set(value, 0);

});

// console.log(config);

/**
 * 此方法是同步方法(不完全是异步方法),这个方法要是都是异步执行就不知道什么时候能执行完了
 * @param  {[type]}  _path [description]
 * @param  {[type]}  tree  =             _tree [description]
 * @return {Boolean}       [description]
 */
function analysisSync(_path, tree = _tree) {
    var files = fs.readdirSync(_path);
    for (var file of files) {
        var stats = fs.statSync(path.join(_path, file));
        if (stats.isFile()) {
            tree.set(file, null);
            analysis_extension(path.join(_path, file));
        } else if (stats.isDirectory()) {
            if (config.excludeDirs.includes(file)) {

            } else if (file.startsWith('.')) {
                // 属于排除的,不扫描的目录
            } else {
                tree.set(file, new Map());

                analysisSync(path.join(_path, file), tree.get(file));
            }
        }
    }

    if (tree == _tree) {
        console.log(extensionCnt);
    }

}

let fileSum = 0;
let dirSum = 0;

function countNumSync(_path,rank=0) {
    var files = fs.readdirSync(_path);
    for (var file of files) {
        var stats = fs.statSync(path.join(_path, file));
        if (stats.isFile()) {
            fileSum++;
        } else if (stats.isDirectory()) {
            if (config.excludeDirs.includes(file)) {
                // 属于排除的,不扫描的目录
            } else if (file.startsWith('.')) {
                // 属于排除的,不扫描的目录
            } else {
                dirSum++;
                countNumSync(path.join(_path, file),rank+1);
            }
        }
    }

    if (rank === 0) {
        console.log('文件的数量为 ',fileSum);
        console.log('文件夹的数量为 ',dirSum);
    }
}

// 统计文件夹,文件的数量
// 还是同样的问题,什么时候知道遍历完了 , 同步的方法会知道什么时候调用完了
function countNum() {
    fs.readdir(_path, function(err, files) {
        if (err) {
            console.log(err);
            console.log('读取文件夹 ', path, ' 出错');
        } else {
            for (var file of files) {
                fs.stat(path.join(_path, file), function(err, stats) {
                    if (stats.isFile()) {
                        fileSum++;
                    } else if (stats.isDirectory()) {
                        if (config.excludeDirs.includes(file)) {
                            // 属于排除的,不扫描的目录
                        } else if (file.startsWith('.')) {
                            // 属于排除的,不扫描的目录
                        } else {
                            dirSum++;
                            analysis(path.join(_path, file), tree.get(file));
                        }
                    }
                });
            }
            console.log(files);
        }
    });
}

/**
 * 此方法是异步方法
 *
 * 一个问题,遍历树形结构的时候,什么时候知道遍历完了
 *
 * @param  {[type]}  _path [description]
 * @param  {[type]}  tree  =             _tree [description]
 * @return {Boolean}       [description]
 */
function analysis(_path, tree = _tree) {

    fs.readdir(_path, function(err, files) {
        if (err) {
            console.log(err);
            console.log('读取文件夹 ', path, ' 出错');
        } else {
            for (var file of files) {
                fs.stat(path.join(_path, file), function(err, stats) {
                    if (stats.isFile()) {
                        tree.set(file, null);
                        analysis_extension(path.join(_path, file));
                    } else if (stats.isDirectory()) {
                        if (config.excludeDirs.includes(file)) {
                            // 属于排除的,不扫描的目录
                        } else if (file.startsWith('.')) {
                            // 属于排除的,不扫描的目录
                        } else {
                            tree.set(file, new Map());
                            // 递归调用
                            analysis(path.join(_path, file), tree.get(file));
                        }
                    }
                });
            }
            console.log(files);
        }
    });
    if (tree == _tree) {
        console.log(extensionCnt);
        //console.log(extensionLineCnt);

    }

}

/**
 * 方法中有异步部分
 * @param  {[type]}  filename [description]
 * @param  {[type]}  extname  [description]
 * @return {Boolean}          [description]
 */
function analysis_line(filename, extname) {
    var data = fs.readFileSync(filename);
    data = data.toString();

    var rl = readline.createInterface({
        input: fs.createReadStream(filename)
    });
    var count = 0;

    // 这个地方是异步的

    rl.on('line', (line) => {
        count++;
    });

    rl.on('close', function() {
        extensionLineCnt.set(extname, extensionLineCnt.get(extname) + count);
    });

}


function analysis_extension(filename) {
    var extname = path.extname(filename);
    if (extname !== '' && config.extensions.includes(extname)) {
        extensionCnt.set(extname, extensionCnt.get(extname) + 1);
        analysis_line(filename, extname);
    }
}

exports.analysis = analysis;
exports.analysisSync = analysisSync;
exports.countNumSync = countNumSync;

File: python/file_analysis.py
import os
import yaml

class FileAnalysis:
    '''
    cofig 是dict,是config.yaml 中的信息
    '''
    def __init__(self, path,config):
      self.basePath = path
      self.config = config
      self.tree={}
      self.extensionCnt={}
      self.extensionLineCnt={}
      self.__initConfig()

    def __initConfig(self):
        for index in range(len(self.config['extensions'])):
            self.extensionCnt[self.config['extensions'][index]]=0
            self.extensionLineCnt[self.config['extensions'][index]]=0



    # 参数的默认值不能使用 self ,应该就是不能使用变量, js的函数默认值是可以使用在前面的参数的
    def analysis(self,path,baseTree=None):
        if baseTree == None:
            baseTree = self.tree
        files=[]
        dirs=[]
        # print('开始分析')
        results = os.listdir(path)
        tempTree={};
        for index in range(len(results)):
            testPath=os.path.join(path,results[index])
            if os.path.isfile(testPath):
                files.append(results[index])
                baseTree[results[index]]=None
                self.__analysis_extension(testPath)
            elif os.path.isdir(testPath):
                if results[index][0]=='.':
                    # 以 . 开头的 暂时认为是排除的文件夹
                    pass
                elif results[index] in self.config['excludeDirs']:
                    # 在排除的目录中
                    pass
                else:
                    dirs.append(results[index])
                    baseTree[results[index]]={}
                    self.analysis(testPath,baseTree[results[index]])

        if baseTree is self.tree:
            result_file=open('result.yaml','w+')
            config = yaml.load(result_file)
            yaml.dump(self.tree,result_file)

    def __analysis_line(self,filename,extension):
        try:
            file = open(filename)
            count = len(file.readlines())
            self.extensionLineCnt[extension]+=count
        except UnicodeDecodeError:
            print('打开 ',filename,' 解码时发生UnicodeDecodeError 异常')
        else:
            file.close()
        # with open(filename) as file:
        #     print(filename)
        #     count = len(file.readlines())
        #     self.extensionLineCnt[extension]+=count

    def __analysis_extension(self,filename):
        result = os.path.splitext(filename)
        # 等于二的才是真正有扩展名的
        if  result[1] != '' and result[1] in self.config['extensions']:
            self.extensionCnt[result[1]]+=1
            self.__analysis_line(filename,result[1])
